### PR TITLE
feat(cli): Add S3StateStore

### PR DESCRIPTION
**Pull request by stbrody**
No automated testing yet as I can't see a good way to test this outside of an integration test that actually speaks to S3, so that will come in https://github.com/ceramicnetwork/js-ceramic/issues/990.  But I have tested manually and everything looked good.

To use this you need to provide the S3 configuration in a way that the AWS library knows to look for it.  I did all my testing by setting the AWS_REGION, AWS_ACCESS_KEY_ID, and AWS_SECRET_ACCESS_KEY environment variables, although I believe there may also be a standard config file location that the AWS library knows to look in, but I didn't test that

### CODE REVIEW

File: packages/core/src/store/level-state-store.ts
Comment (valmack) [inline packages/core/src/store/level-state-store.ts:30]: Can we add some error handling here?

For example if the process doesn't have write access to the given path, it'd be helpful to return a descriptive error and im guessing we want to exit in that case.

File: packages/core/src/store/pin-store-factory.ts
Comment (valmack) [inline packages/core/src/store/pin-store-factory.ts:39]: Seems like this method name should be `create`? Bc it creates and returns a new StateStore

File: packages/core/src/index.ts
Comment (valmack) [inline packages/core/src/index.ts:3]: What do we gain from this export?

File: packages/core/src/store/pin-store.ts
Comment (valmack) [inline packages/core/src/store/pin-store.ts:18]: So the "StateStore" holds both the state of any document commits/tips loaded and the tips of any documents created. Does that all happen through the PinStore?

Here it seems there could be other entities besides the PinStore that could use the same stateStore, however the `open` method presumably should be called only once so it seems out of place here.

If my understanding is correct my suggestion would be to open the stateStore in the Ceramic initialization instead.

File: packages/core/src/store/level-state-store.ts
Comment (stbrody) [reply packages/core/src/store/level-state-store.ts:30]: If this fails for any reason it throws an error which bubbles up to the top level.  As an experiment I tried deleting the `statestore` directory and taking away write access from the `~/.ceramic` directory and ran the daemon.  I got an error: 
`Error: EACCES: permission denied, mkdir '/home/spencer/.ceramic/statestore/testnet-clay'`.  That kinda seems good enough for now?

File: packages/cli/src/bin/ceramic.ts
Comment (stbrody) [reply packages/cli/src/bin/ceramic.ts:12]: Renamed to log-directory

File: packages/core/src/index.ts
Comment (stbrody) [reply packages/core/src/index.ts:3]: The ability to import the `StateStore` interface in the `cli` package, where the `S3StateStore` is defined

File: packages/core/src/store/pin-store-factory.ts
Comment (stbrody) [reply packages/core/src/store/pin-store-factory.ts]: renamed to `createPinStore()`

File: packages/core/src/store/pin-store.ts
Comment (stbrody) [reply packages/core/src/store/pin-store.ts:18]: hmm, the `StateStore` *is* sometimes used directly, but always via a handle from the pinStore.  I can't imagine there ever being a time when we would have a `StateStore` but not have a PinStore.  In fact, with the stuff Sergey is working on now, I expect all these internal pieces to become even more hidden away behind the new `DocStateRepository` interface.  So I think we want to be moving in the direction of pushing more of these pieces down behind interfaces rather than pulling more of them up to be used directly in higher layers

**Comment by oed** [comment]
> To use this you need to provide the S3 configuration in a way that the AWS library knows to look for it. I did all my testing by setting the AWS_REGION, AWS_ACCESS_KEY_ID, and AWS_SECRET_ACCESS_KEY environment variables, although I believe there may also be a standard config file location that the AWS library knows to look in, but I didn't test that

Is there no way to pass this to the library on construction time? Would be nice if we could set this configuration in `~/.ceramic/config.json`.

**Comment by stbrody** [comment]
> > To use this you need to provide the S3 configuration in a way that the AWS library knows to look for it. I did all my testing by setting the AWS_REGION, AWS_ACCESS_KEY_ID, and AWS_SECRET_ACCESS_KEY environment variables, although I believe there may also be a standard config file location that the AWS library knows to look in, but I didn't test that
> 
> Is there no way to pass this to the library on construction time? Would be nice if we could set this configuration in `~/.ceramic/config.json`.

Yes we could.  In fact, my first version of this PR did it that way.  Take a look at [this commit](https://github.com/ceramicnetwork/js-ceramic/pull/1041/commits/8f52adb4436bdba84da39718a04ef3df44ae0a9a) from this PR's commit history.  I could revert that commit and it would be possible to explicitly configure these settings.  The reason I ripped that out is that I came to understand that the AWS package is already configured to look for aws configuration options in certain well known places, so if we don't provide the options explicitly, it will load them via the standard AWS ways, which includes its [own config file|https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-region.html#setting-region-config-file].  So I opted to leave things to work in the "standard" way for AWS configs.

I have no idea though how other libraries that utilize the AWS api tend to expose these configs - whether its actually common to let the inner AWS library handle the configuration in the way it wants to, or if it's most common to repackage those config options as configs to the higher-level library.

**Comment by stbrody** [comment]
> Would be nice if we could set this configuration in ~/.ceramic/config.json

Just FYI, the `config.json` file currently can only be used to set options for the http-client started by the CLI when interacting *with* a daemon for creating documents, etc.  It is not used when *starting* the daemon.  If we want a way for a config file to control the actual full ceramic node that is started when running `ceramic daemon` we'll need to do more work to enable that.

File: packages/core/src/store/pin-store.ts
Comment (valmack) [reply packages/core/src/store/pin-store.ts:18]: Ok so state store is only used if you explicitly choose to pin something? If that is the case then it makes sense how it is

**Comment by valmack** [comment]
> If we want a way for a config file to control the actual full ceramic node that is started when running `ceramic daemon` we'll need to do more work to enable that.

it’s definitely helpful to have be forced to set config from two different places

As long as we keep ability to use env vars I’m happy.

File: packages/cli/src/bin/ceramic.ts
Comment (valmack) [reply packages/cli/src/bin/ceramic.ts:12]: Oof I should have said, path makes more sense because we need full path, not just directory name.

**Comment by oed** [comment]
> Yes we could. In fact, my first version of this PR did it that way. Take a look at this commit from this PR's commit history. I could revert that commit and it would be possible to explicitly configure these settings. The reason I ripped that out is that I came to understand that the AWS package is already configured to look for aws configuration options in certain well known places, so if we don't provide the options explicitly, it will load them via the standard AWS ways, which includes its own config file. So I opted to leave things to work in the "standard" way for AWS configs.

> I have no idea though how other libraries that utilize the AWS api tend to expose these configs - whether its actually common to let the inner AWS library handle the configuration in the way it wants to, or if it's most common to repackage those config options as configs to the higher-level library.

Ehm, do we document this anywhere in that case? Seems kind of bad that you need to know how an underlaying library works in order to use this feature in ceramic?

**Comment by stbrody** [comment]
> > Yes we could. In fact, my first version of this PR did it that way. Take a look at this commit from this PR's commit history. I could revert that commit and it would be possible to explicitly configure these settings. The reason I ripped that out is that I came to understand that the AWS package is already configured to look for aws configuration options in certain well known places, so if we don't provide the options explicitly, it will load them via the standard AWS ways, which includes its own config file. So I opted to leave things to work in the "standard" way for AWS configs.
> 
> > I have no idea though how other libraries that utilize the AWS api tend to expose these configs - whether its actually common to let the inner AWS library handle the configuration in the way it wants to, or if it's most common to repackage those config options as configs to the higher-level library.
> 
> Ehm, do we document this anywhere in that case? Seems kind of bad that you need to know how an underlaying library works in order to use this feature in ceramic?

Yep, that's a totally fair concern.  I'm happy to revert that commit and put these options into explicit config settings for the CeramicDaemon if you think that's best.  I just wanted to raise the issue that there does seem to be some sort of standard way to provide this configuration.  Also if there are other client-side AWS options that users might want to control, we'd have to expose each one ourselves in that route, vs if we let the AWS library handle it then users have access to all the underlying config options they might want.  We will certainly have to make sure to document this, but we'd have to have some docs of these options in either case anyway.

File: packages/cli/src/bin/ceramic.ts
Comment (stbrody) [reply packages/cli/src/bin/ceramic.ts:12]: I guess I went with directory over path b/c in my mind "path" sounds like it's the full path to the specific file where the logs or state store will be, when in fact it's just the path to the directory where we will create files and possibly sub-directories according to a layout we decide.

File: packages/core/src/store/pin-store.ts
Comment (stbrody) [reply packages/core/src/store/pin-store.ts:18]: Yeah, there will only be state in the state store for docs that are pinned

Comment (valmack) [reply packages/core/src/store/pin-store.ts:18]: 🙏 thanks for clarity

File: packages/cli/src/bin/ceramic.ts
Comment (valmack) [reply packages/cli/src/bin/ceramic.ts:12]: Either way consistency is most important thing so this is good :)

**Review by valmack — APPROVED** [review]
👍

**Comment by zachferland** [comment]
yeah its very common and i think expected really, for most all libraries to leave that to underlying aws sdk, and you may configure other options there as well. As spencer mentioned it already pull these vars from some default places and this allows aws sdk to make assumptions about where its going to fetch them during runtime depending on the aws service it is running in (you wont necessarily be able to easily fetch these values ahead of runtime in some service configurations).

Also for testing, may make sense to test locally with something like https://github.com/minio/minio, object store with same s3 interface.

**Comment by stbrody** [comment]
> Also for testing, may make sense to test locally with something like https://github.com/minio/minio, object store with same s3 interface.

Interesting.  You're suggesting starting an in-process minio instance in the jest test, and pointing the ceramic node at that for its S3 storage?  When I looked briefly at minio it looked much more heavyweight than that, like it can actually be used in production deployments rather than just within tests.  But having some kind of lightweight S3-api compatible service we could start within the tests would certainly be really great.  But so far I can't even find a way to start a minio server from within javascript?  Do you know if this is possible @zachferland?

**Comment by zachferland** [comment]
yeah agree on second look again, likely too heavy for tests here still, although i thought that project/community also had some test/mock tooling. Just in 3box before we ran a docker image with it in local configurations, local development and for some more integration style tests.

**Comment by stbrody** [comment]
@oed, okay to merge as-is with config management delegated to AWS library?

**Review by oed — APPROVED** [review]

**Comment by nahsi** [comment]
> > Also for testing, may make sense to test locally with something like https://github.com/minio/minio, object store with same s3 interface.
> 
> Interesting. You're suggesting starting an in-process minio instance in the jest test, and pointing the ceramic node at that for its S3 storage? When I looked briefly at minio it looked much more heavyweight than that, like it can actually be used in production deployments rather than just within tests. But having some kind of lightweight S3-api compatible service we could start within the tests would certainly be really great. But so far I can't even find a way to start a minio server from within javascript? Do you know if this is possible @zachferland?

Usually you start a development instance of minio in docker and then use it to run tests against.

There is a library for that for golang https://github.com/testcontainers/testcontainers-go . For javascripts something similar should exist.

**Pull request merged**
